### DISCUSSION TITLE
PlugData extensions (developing and using)

### DISCUSSION THREAD

**Discussion by zepadovani** [plugdata-team/plugdata#186] [General]
Hi,

first of all, congratulations for the project: very promising! I'm using with my "electroacoustic music" and "interactive music systems" students and it's proving to be a really a nice gateway to computer music (specially for students more used to using DAWs).

I have some questions about extensions:

1. I see that the extensions available for download do not include all of those in Deken. For example, I have made a set of abstractions called "planifolia" (a pack of vanilla abstractions) that is available in deken, but I cannot find it in PlugData. Thus, I would like to know how is the process to follow to publish new extensions in PlugData (including those that require the compilation of externals, since we are developing some objects here).
2. Are there any relevant precautions when developing objects (in C or C++) that I should consider so that they are compatible with both vanilla PD and PlugData?
3. Some externals run without problems in the standalone version but post errors while I try to load them in the lv2 plugin (PlugDataFX). iemlib, for instance, loads correctly in the standalone but prints this in the plugin:
`/home/padovani/.config/PlugData/Library/Deken/iemlib/iemlib.pd_linux: /home/padovani/.config/PlugData/Library/Deken/iemlib/iemlib.pd_linux: undefined symbol: s_list`

Is this something related to the architecture of lv2 plugins? (I ask because I'm not sure if it is an 'iemlib' or a 'PlugData' bug)

Thanks!

**Comment by timothyschoen** [comment]
Thanks! Awesome to hear you're using it with your students!

1. If you're external isn't showing up in Deken, that's probably my fault. PlugData gets the deken information from [here](https://github.com/timothyschoen/PlugDataDekenServer/tree/main), which is a simple repo that updates daily, and preprocesses the available Deken packages to a format that's easy to digest for PlugData. I'll find out why yours isn't showing up! 
Just in case: are you using M1 Mac? Then you might want to run PlugData in Rosetta.

2. For object compatibility: PlugData can't render custom Tcl/Tk graphics yet. Other than that, it should work the same. At some point, I'd like to create a way for externals to create their own graphics.

3.  That's a PlugData/libpd problem, see this: https://github.com/timothyschoen/PlugData/issues/34
I've tried to fix it but it's very tricky. Plugins need to be able to load multiple separate instances of pure-data (for thread-safety), which changes interface that externals use to communicate with Pd.

**Reply by zepadovani** [reply]
Hi!

Not sure if the comment here (https://github.com/timothyschoen/PlugData/issues/34#issuecomment-1186720205) means that, using your workaround, externals can be compiled to work properly as PlugDataFX externals, as PlugData (standalone) externals and as PD vanillas externals. If it is so, please consider writing some sort of short step-by-step guide on how to change the sources to do this! Even if it means to recompile externals (or even PD vanilla), it would be very usefull!

BTW: I'm using mainly Manjaro x64 now...

Thank you again!

**Reply by timothyschoen** [reply]
Agreed! It is possible to compile externals specifically for PlugData's plugin version, but I've found that some externals don't compile at all when you do this... I'll add a wiki section about how to do this later today!

**Reply by zepadovani** [reply]
Wonderful! Thank you very much!!

**Reply by timothyschoen** [reply]
I see that planifolia is just abstractions, so it should suffice to put it in PlugData's search path. On Linux, that would be ~/.config/PlugData/Library/Abstractions (or ~/.config/PlugData/Deken, both are okay).

This will also work in the plugin, the restriction is only for C-coded externals.

I'll also add a tutorial for C-coded ones in the Wiki though, because that is very useful.

**Reply by timothyschoen** [reply]
Okay, I see now why it didn't compile: this will probably only work on Linux. On Mac (and I think Windows as well), thread-local storage works differently inside a dynamic library...

I'll write a wiki page about this later, it's a bit more complicated that I had hoped.

**Reply by timothyschoen** [reply]
<img width="800" alt="Screenshot 2022-09-12 at 04 50 34" src="https://user-images.githubusercontent.com/44585538/189565256-09bd24ef-dca9-438c-a24a-b19859a5f1c2.png">

It shows up in the package manager now!

**Reply by zepadovani** [reply]
nice!! thank you!
